### PR TITLE
(Draft) Use OgreThreadDefines.h for threading

### LINKED ISSUE #13: Use std::thread instead of boost::thread?
Hello @paroj, I'm trying to replace the use of `boost::thread` with that of `std::thread` and I want to ask you some questions.

On one hand if you think that `boost::thread` should remain as a third option or be replaced by `std::thread` entirely.

Also, can you confirm that this replacement is valid?
```
//boost::recursive_mutex::scoped_lock soundLock(mSoundMutex);
std::lock_guard<std::recursive_mutex> soundLock(mSoundMutex);
```

Another thing is that I'm trying to find a replacement for this statement, but it goes over my head:
```
mUpdateThread = OGRE_NEW_T(boost::thread, Ogre::MEMCATEGORY_GENERAL)(boost::function0<void>(&OgreOggSoundManager::threadUpdate, this));
```

Can you help?

Thanks.

> **NOTE:** If you want I can submit the pull request and you review there. But I wasn't able to test anything because for now the code is not compiling due to the mUpdateThread creation.

### PR DESCRIPTION
**Pull request by sercero**
fixes #13

### CODE REVIEW

File: oggsound/src/OgreOggSoundManager.cpp [hunk 1]
```diff
@@ -398,7 +406,13 @@ namespace OgreOggSound
 #	else
 		mUpdateThread = OGRE_NEW_T(boost::thread, Ogre::MEMCATEGORY_GENERAL)(boost::function0<void>(&OgreOggSoundManager::threadUpdate, this));
 		Ogre::LogManager::getSingleton().logMessage("*** --- Using BOOST threads for streaming", Ogre::LML_NORMAL);
-#	endif	
+#	endif
+*/
+		#define OGRE_THREAD_CREATE(name, worker) boost::thread* name = OGRE_NEW_T(boost::thread, MEMCATEGORY_GENERAL)(worker)
```
Comment (paroj) [inline oggsound/src/OgreOggSoundManager.cpp:411]: this

File: oggsound/include/OgreOggSoundManager.h [hunk 2]
```diff
@@ -638,6 +642,10 @@ namespace OgreOggSound
 		static boost::recursive_mutex mSoundMutex;
 		static boost::recursive_mutex mResourceGroupNameMutex;
 #	endif
+*/
+		OGRE_WQ_MUTEX(mMutex);
```
Comment (paroj) [inline oggsound/include/OgreOggSoundManager.h:646]: we must use the _WQ_ variants as the other ones are noops with OGRE_THREAD_SUPPORT=3

Comment (sercero) [reply oggsound/include/OgreOggSoundManager.h:646]: Oh, I see. Thanks

Comment (sercero) [reply oggsound/include/OgreOggSoundManager.h:646]: @paroj I pushed a new version with your fixes, now the code compiles but I don't know how well it works. Haven't been able to test it yet.

Are you able to test this with POCO, Boost and TBB?

This is only for streaming right?

File: oggsound/include/OgreOggSoundManager.h [hunk 3]
```diff
@@ -722,19 +733,26 @@ namespace OgreOggSound
...
 			{	
 				{
+/*
 #ifdef POCO_THREAD
 					Poco::Mutex::ScopedLock l(OgreOggSoundManager::getSingletonPtr()->mMutex);
 #else
 					boost::recursive_mutex::scoped_lock lock(OgreOggSoundManager::getSingletonPtr()->mMutex);
 #endif
+*/
+					OGRE_LOCK_MUTEX_NAMED(OgreOggSoundManager::getSingletonPtr()->mMutex, lock);
```
Comment (paroj) [inline oggsound/include/OgreOggSoundManager.h:743]: these also need to be
```suggestion
					OGRE_WQ_LOCK_MUTEX_NAMED(OgreOggSoundManager::getSingletonPtr()->mMutex, lock);
```

File: oggsound/src/OgreOggListener.cpp [hunk 4]
```diff
@@ -36,23 +36,29 @@
...
-#if OGGSOUND_THREADED
+#if OGRE_THREAD_PROVIDER
+/*
 #	ifdef POCO_THREAD
 	Poco::Mutex OgreOggSound::OgreOggListener::mMutex;
 #	else
 	boost::recursive_mutex OgreOggSound::OgreOggListener::mMutex;
 #	endif
+*/
+	OGRE_STATIC_MUTEX_INSTANCE(OgreOggSound::OgreOggListener::mMutex);
```
Comment (paroj) [inline oggsound/src/OgreOggListener.cpp:47]:
```suggestion
	OGRE_WQ_MUTEX(OgreOggSound::OgreOggListener::mMutex);
```

File: oggsound/include/OgreOggSoundManager.h (hunk 2 above)
Comment (paroj) [reply oggsound/include/OgreOggSoundManager.h:646]: no idea.. never used threading here

Comment (paroj) [reply oggsound/include/OgreOggSoundManager.h:646]: > Are you able to test this with POCO, Boost and TBB?

I would just leave it after porting to ogre threading. We can then wait until someone with an actual application using this comes along to properly test it.

Comment (sercero) [reply oggsound/include/OgreOggSoundManager.h:646]: OK

**Comment by sercero** [comment]
@paroj I need to know the value of the variable `OGRE_CONFIG_THREAD_PROVIDER` from within the Ogre Audio-Video CMakelists.txt, how can I do that?

That is because I need to know which library to include (boost/POCO/TBB/etc...)

Thanks!

**Comment by paroj** [comment]
it should be sufficient to link against OgreMain as this cmake target will implicitly pull in any threading libraries it uses

File: oggsound/src/OgreOggListener.cpp (hunk 4 above)
Comment (paroj) [inline oggsound/src/OgreOggListener.cpp:47]: this is a noop. You can safely delete OGRE_STATIC_MUTEX_INSTANCE everywhere

File: oggsound/src/OgreOggListener.cpp [hunk 5]
```diff
@@ -36,23 +36,29 @@
...
 #endif
 
 	/*/////////////////////////////////////////////////////////////////*/
 	void OgreOggListener::setPosition(ALfloat x, ALfloat y, ALfloat z)
 	{
-#if OGGSOUND_THREADED
+#if OGRE_THREAD_PROVIDER
```
Comment (paroj) [inline oggsound/src/OgreOggListener.cpp:53]: for merging you should delete any commented code and redundant #if

here, OGRE_LOCK_MUTEX_NAMED will be a noop if OGRE_THREAD_PROVIDER == 0

**Comment by paroj** [comment]
do you have any issues with the ogre_threads branch?

**Comment by sercero** [comment]
Yes, it has the same issue as my branch `std_thread` (I did not submit that PR).

But the crash is exactly the same as #16 

It crashes when doing `mUpdateThread->join();`.

To trigger this you have to be running inside a debugger, otherwise the program exists normally both with `Debug` and `Release`   versions.

**Comment by paroj** [comment]
hmm.. seems this is platform specific. I cannot reproduce this on Linux with the player demo - even with a debugger attached.

**Comment by sercero** [comment]
I see, might be something with Windows + MinGW64.

So it's important to keep the flag `OGGSOUND_THREADED` for people that might hit this bug and want to disable threading as a result.

In my case, I wasn´t able to find a solution, tried many things but eventually gave up.

Perhaps someday I will try again, for now I want to do some other things.

Thanks!

**Pull request closed without merging**